src/cljam/algo/pileup.clj

```diff
@@ -142,6 +115,159 @@
+                            [b1 _] (seqs1 relative-pos1)
+                            [b2 _] (seqs2 relative-pos2)]
+                         (if (= b1 b2)
+                           [(min 200 (+ q1 q2)) 0]
+                           (if (<= q2 q1)
+                             [(int (* 0.8 q1)) 0]
+                             [0 (int (* 0.8 q2))]))))
+          [new1 new2] (apply map vector new-quals)]
+      {(keyword (str tlen1))
+       (merge-corrected-quals a1 correct-start new1)
+       (keyword (str tlen2))
+       (merge-corrected-quals a2 correct-start new2)})))
+
+(defn- make-corrected-quals-map
+  "Make a map which has corrected quals of all overlapping pairs."
+  [alns]
+  (->> alns
+       (group-by (fn [x] (.qname ^SAMAlignment x)))
+       (into {} (map
+                 (fn [[qname xs]]
+                   (if (<= (count xs) 1)
+                     {}
+                     {(keyword qname) (apply correct-pair-quals xs)}))))))
+
+(defn- correct-quals-at-ref
+  "Returns an alignment with corrected quals by looking up quals
+  in the corrected quals map."
+  [corrected-map ^SAMAlignment aln]
+  (let [qname (.qname aln)
+        tlen  (.tlen aln)]
+    (if (zero? tlen)
+      aln
+      (if-let [new-quals (get-in corrected-map
+                                 [(keyword qname) (keyword (str tlen))])]
+        (assoc aln :quals-at-ref new-quals)
+        aln))))
+
+(defn- get-max-pnext
+  "Get the maximum pnext of alignments."
+  [alns]
+  (->> alns
+       (map (fn [^SAMAlignment aln]
+              (if (= (.rnext aln) "=")
+                (.pnext aln)
+                0)))
+       (apply max)))
+
+(defn- read-chunk
+  "Read a chunk which is specified by the region."
+  [sam-reader min-map-quality region]
+  (->> region
+       (sam/read-alignments sam-reader)
+       (sequence
+         (comp
+           (filter (basic-mpileup-pred min-map-quality))
+           (map index-cigar)))))
+
+(defn- get-chunk
+  "Get a chunk in which all pairs exist.
+  Returns the end position of the chunk and alignments in the chunk.
+  It concatenates chunks when a pair overlap across multiple chunks."
+  [sam-reader min-map-quality last-end {:keys [chr start end] :as region}]
+  (loop [s    start
+         e    end
+         alns (read-chunk sam-reader min-map-quality region)]
+    (if (or (empty? alns)                ;; empty chunk
+            (= e last-end)               ;; this is the last chunk
+            (<= (get-max-pnext alns) e)) ;; all pairs are in this chunk
+      [e alns]
+      (let [next-start (inc e)
+            next-end   (min last-end (+ e (- e s) 1))]
+        (recur
+          next-start
+          next-end
+          (->> (read-chunk sam-reader
+                           min-map-quality
+                           {:chr   chr
+                            :start next-start
+                            :end   next-end})
+               (concat alns)
+               distinct))))))
```

Comment:
`distinct` and multiple `seek`s might be too costly.
Just a suggestion, but could you try chunking like the following?

```clojure
;; - it's almost the same as `pileup-seq`, could be merged.
;; - need some unit tests
(defn chunk-seq
  [^long start ^long end ^long chunk-size alns]
  (letfn [(step [^long pos prev-buf rest-alns]
            (lazy-seq
             (when (< (dec pos) end)
               (let [[i rests] (split-with (partial cover-locus? (dec (+ pos chunk-size))) rest-alns)
                     p (some-> rests ^SAMAlignment first .pos)
                     [b :as buf] (into prev-buf i)
                     [nb :as next-buf] (filterv #(<= (+ pos chunk-size) (.end ^SAMAlignment %)) buf)
                     next-pos (if (and p (not nb)) p (+ pos chunk-size))]
                 (if b
                   (cons [pos buf] (step next-pos next-buf rests))
                   (when (and p (<= p end))
                     (step next-pos next-buf rests)))))))]
    (let [[x :as xs] (drop-while #(< (.end ^SAMAlignment %) start) alns)]
      (when x
        (step (max start (.pos ^SAMAlignment x)) [] xs)))))
```

```clojure
(defn pileup-chunks
  [sam-reader {:keys [chr start end] :as region} chunk-size
   min-base-quality min-map-quality ignore-overlaps?]
  (->> (sam/read-alignments sam-reader region)
       (sequence
        (comp
         (filter (basic-mpileup-pred min-map-quality))
         (map index-cigar)))
       (chunk-seq start end chunk-size)
       (sequence
        (comp
         (mapcat (fn [[pos alns]]
                   (->> (if ignore-overlaps?
                          alns
                          (keep (partial correct-quals-at-ref
                                         (make-corrected-quals-map alns)) alns))
                        (pileup-seq pos (min end (dec (+ pos chunk-size)))))))
         (map resolve-bases)
         (if (pos? min-base-quality)
           (map (filter-by-base-quality min-base-quality))
           identity)
         (keep (partial ->locus-pile chr))))))
```

Reply:
Thank you for the great suggestion!
I've tried and measured the performance.
It improved the performance for long-region regions.

- sample: WES SRP006438.sorted.bam
  - many alignments chr1:152210000-152214999  

    |branch|time|
    |----|----|
    |master|time: 10.852575 sec, sd: 2.435336 ms|
    |(previous) feature/pileup-chunks|time: 7.804254 sec, sd: 53.081289 ms|
    |(new) feature/pileup-chunks|time: 7.640314 sec, sd: 2.162599 ms|

  - gene: NRAS chr1:114703000-114723000
    
    |branch|time|
    |----|----|
    |master|time: 2.424547 sec, sd: 14.857820 µs|
    |(previous) feature/pileup-chunks|time: 2.205339 sec, sd: 7.169653 µs|
    |(new) feature/pileup-chunks|time: 1.700564 sec, sd: 3.104096 µs|

  - gene: EGFR chr7:55015000-55215000
    
    |branch|time|
    |----|----|
    |master|time: 5.902383 sec, sd: 30.402538 µs|
    |(previous) feature/pileup-chunks|time: 5.498727 sec, sd: 54.685996 µs|
    |(new) feature/pileup-chunks|time: 4.197941 sec, sd: 249.036283 µs|